src/main/io/osd.c

```diff
@@ -425,19 +428,34 @@ static void osdDrawSingleElement(uint8_t item)
 #endif // GPS
 
     case OSD_COMPASS_BAR:
-        {
-            int16_t h = DECIDEGREES_TO_DEGREES(attitude.values.yaw);
-
-            h = osdGetHeadingIntoDiscreteDirections(h, 16);
+    {
+#ifdef USE_GPS
+        int16_t h = 0;
+		// If MAG_Heading is not available, yaw starts with zero in a random direction. With GPS course over ground the flight direction is roughly displayed instead
+		if (STATE(GPS_FIX) && STATE(GPS_FIX_HOME) && !FLIGHT_MODE(MAG_MODE)) {
+			h = DECIDEGREES_TO_DEGREES(gpsSol.groundCourse);
+		}
+#else
+        int16_t h = DECIDEGREES_TO_DEGREES(attitude.values.yaw);
+#endif
+        h = osdGetHeadingIntoDiscreteDirections(h, 16);
 
-            memcpy(buff, compassBar + h, 9);
-            buff[9]=0;
-            break;
-        }
+        memcpy(buff, compassBar + h, 9);
+        buff[9]=0;
+        break;
+    }
 
     case OSD_ALTITUDE:
         {
+#ifdef USE_GPS
+			// If BARO is not available, the estimated altitude is zero. Using GPS altitude an absolute height can be displayed instead.
+			if (STATE(GPS_FIX) && !FLIGHT_MODE(BARO_MODE)) {
```

Comment:
I agree with @DanNixon that using GPS altitude here and estimated altitude everywhere else confusing. GPS altitude should be used to calculate the estimated altitude in `calculateEstimatedAltitude`.

Reply:
In principle i agree but this would need a sophisticated sensor fusion / EKF.
Here i wanted a small solution if no other altitude sensor (BARO or RANGE_FINDER) is available. I sharpened the condition.
My experience was that the altiude estimation merely based on accelerometer integration is zero.